cedar-drt/fuzz/src/lib.rs

```diff
@@ -177,47 +177,53 @@ pub fn run_val_test(
 
     let definitional_res = custom_impl.validate(&schema, policies, mode);
 
-    match definitional_res {
-        Err(err) => {
-            // TODO(#175): For now, ignore cases where the Lean code returned an error due to
-            // an unknown extension function.
-            if !err.contains("jsonToExtFun: unknown extension function") {
-                panic!(
-                    "Unexpected parse error\nPolicies:\n{}\nSchema:\n{:?}\nError: {err}",
-                    &policies, schema
-                );
-            }
-        }
-        Ok(definitional_res) => {
-            // TODO:(#126) Temporary fix to ignore a known mismatch between the spec and `cedar-policy`.
-            // The issue is that the `cedar-policy` will always return an error for an
-            // unrecognized entity or action, even if that part of the expression
-            // should be excluded from typechecking (e.g., `true || Undefined::"foo"`
-            // should be well typed due to short-circuiting).
-            if rust_res.validation_errors().any(|e| {
-                matches!(
-                    e.error_kind(),
-                    ValidationErrorKind::UnrecognizedEntityType(_)
-                        | ValidationErrorKind::UnrecognizedActionId(_)
-                )
-            }) {
-                return;
+    // If `cedar-policy` does not return an error, then the spec should not return an error.
+    // This implies type soundness of the `cedar-policy` validator since type soundness of the
+    // spec is formally proven.
+    //
+    // In particular, we have proven that if the spec validator does not return an error (B),
+    // then there are no authorization-time errors modulo some restrictions (C). So (B) ==> (C).
+    // DRT checks that if the `cedar-policy` validator does not return an error (A), then neither
+    // does the spec validator (B). So (A) ==> (B). By transitivity then, (A) ==> (C).
+
+    if rust_res.validation_passed() {
+        match definitional_res {
+            Err(err) => {
+                // TODO(#175): For now, ignore cases where the Lean code returned an error due to
+                // an unknown extension function.
+                if !err.contains("jsonToExtFun: unknown extension function") {
+                    panic!(
+                        "Unexpected parse error\nPolicies:\n{}\nSchema:\n{:?}\nError: {err}",
+                        &policies, schema
+                    );
+                }
             }
+            Ok(definitional_res) => {
+                // Even if the Rust validator succeeds, the definitional validator may
+                // return "impossiblePolicy" due to greater precision. In this case, the
+                // input policy is well-typed, although it is guaranteed to always evaluate
+                // to false.
+                if definitional_res
+                    .validation_errors
+                    .contains(&"impossiblePolicy".to_string())
```

Comment:
what happens here if the definitional validator returns multiple errors, one of which is impossiblePolicy?  Or does the definitional validator never return multiple errors?

Reply:
The definitional validator currently always returns a single error. But I'll switch this code to check that the list has length 1 in case we change that in the future.